### PR TITLE
[C++ VERIFICATION] Improved cpp new and delete

### PR DESCRIPTION
**Pull request by XLiZHI**
Fixed C++ dangling pointer checking. 

Fixed C++ double delete.

Next PR todo: move the adjustment of cpp new and delete into symex.

still wait for #1644

### CODE REVIEW

File: regression/esbmc-cpp/cpp/ dangling_ptr/test.desc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.cpp
+
+^VERIFICATION SUCCESSFUL$
```
Comment (lucasccordeiro) [inline regression/esbmc-cpp/cpp/ dangling_ptr/test.desc]: There is a memory leak in this program.

Can you please create another test case that enables --memory-leak-leak?

File: regression/esbmc-cpp/cpp/ dangling_ptr/main.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+int main()
+{
+  Foo *foo = new Foo();
+
+  foo->Increment();  // Incrementing the value
+
+  foo->Execute(); // Executing
+
```
Comment (lucasccordeiro) [inline regression/esbmc-cpp/cpp/ dangling_ptr/main.cpp]: Can you also create another test case that adds the delete to mark it as VERIFICATION SUCCESSFUL?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
I have two comments here.

**Review by lucasccordeiro — APPROVED** [review]

**Review by fbrausse — APPROVED** [review]
You can unify both branches by casting `*code2` to the common base class `const code_expression_data &`. Otherwise LGTM.

**Comment by XLiZHI** [comment]
> You can unify both branches by casting `*code2` to the common base class `const code_expression_data &`. Otherwise LGTM.

Hi @fbrausse,

Thanks for your advice, I tried this, but cmake complained about it.
```
const code_expression_data &deref_code = is_code_cpp_delete2t(code2)? to_code_cpp_delete2t(code2)
                                                                    : to_code_cpp_del_array2t(code2);
```
different types ‘code_cpp_delete2t’ and ‘code_cpp_del_array2t’. Or do we have a general type cast function?

**Comment by fbrausse** [comment]
`const auto &code_expr = static_cast<const code_expression_data &>(*code2);` should work.

**Comment by XLiZHI** [comment]
> const auto &code_expr = static_cast<const code_expression_data &>(*code2);

Great, thanks

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @XLiZHI.

**Pull request merged**